### PR TITLE
86 create libcalapi services for libcal

### PR DESCRIPTION
**Pull request by TranHVLoc**
**Utility** and **cancel reservation** service have been done.
Try pushing pull request to check for code Nest code convention from @dominhnhut01. JUnit test haven't been written. 

After this pull request is approved, the branch will be deleted and for any other service created, there will be the following issue to specify that pull request for a certain service only

### CODE REVIEW

File: new-chatbot-app/src/library-api/library-api.module.ts [hunk 1]
```diff
@@ -1,12 +1,17 @@
...
 
 @Module({
-  imports: [HttpModule, SharedModule],
-  providers: [LibappsAuthorizationService],
-  exports: [LibappsAuthorizationService],
+  imports: [SharedModule],
+  providers: [
+    CancelReservationService,
+    LibcalApiAuthorizationService,
+    PrismaService,
```
Comment (dominhnhut01) [inline new-chatbot-app/src/library-api/library-api.module.ts:13]: PrismaService shouldn't belong to LibraryApiModule

File: new-chatbot-app/src/llm-chain/llm-toolbox/libcal-tools/cancel-reservation/cancel-reservation.service.ts [hunk 2]
```diff
@@ -0,0 +1,82 @@
+import { Injectable } from '@nestjs/common';
+import { NetworkService } from 'src/shared/services/network/network.service';
+import { Room, LibcalApiAuthorizationService } from '../../../../library-api/libcal-api-auth/libcal-api-auth.service';
+
+@Injectable()
+export class CancelReservationService {
```
Comment (dominhnhut01) [inline new-chatbot-app/src/llm-chain/llm-toolbox/libcal-tools/cancel-reservation/cancel-reservation.service.ts:6]: This class should implement interface LlmTool. And I suggest changing the name to CancelReservationToolService to fit our current naming strategy

File: new-chatbot-app/src/llm-chain/llm-toolbox/libcal-tools/cancel-reservation/cancel-reservation.service.ts [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+   */
+  async serviceRun(toolInput: { bookingID: string | null }): Promise<string> {
+    return new Promise<string>(async (resolve, reject) => {
+      if (
+        toolInput.bookingID === null ||
+        toolInput.bookingID === "null" ||
+        toolInput.bookingID === undefined ||
+        toolInput.bookingID === "undefined"
+      ) {
+        console.log(
```
Comment (dominhnhut01) [inline new-chatbot-app/src/llm-chain/llm-toolbox/libcal-tools/cancel-reservation/cancel-reservation.service.ts:37]: We shouldn't use console.log. It should only be used during debug. Reference: https://stackoverflow.com/questions/8002116/should-i-be-removing-console-log-from-production-code
If you want to log, I suggest using the Logging framework of Nest.js instead (https://www.coreycleary.me/should-you-use-a-logging-framework-or-console-log-in-node/)

File: new-chatbot-app/src/llm-chain/llm-toolbox/libcal-tools/cancel-reservation/cancel-reservation.service.ts [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    private libcalApiAuthService: LibcalApiAuthorizationService
+  ) {}
+
+  /**
+   * Run the cancel reservation service.
+   * @param toolInput 
+   * @returns
+   */
+  async serviceRun(toolInput: { bookingID: string | null }): Promise<string> {
+    return new Promise<string>(async (resolve, reject) => {
```
Comment (dominhnhut01) [inline new-chatbot-app/src/llm-chain/llm-toolbox/libcal-tools/cancel-reservation/cancel-reservation.service.ts:30]: new version of Typescript doesn't require this anymore. I suggest remove this and simply return instead of resolve

File: new-chatbot-app/src/llm-chain/llm-toolbox/libcal-tools/cancel-reservation/cancel-reservation.service.ts [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    return new Promise<{ success: boolean; error: string | null }>(
+      async (resolve, reject) => {
+        const accessToken = await this.libcalApiAuthService.getAccessToken();
+        const header = {
+          Authorization: `Bearer ${accessToken}`
+        };
+      }
+    );
+  }
+}
```
Comment (dominhnhut01) [inline new-chatbot-app/src/llm-chain/llm-toolbox/libcal-tools/cancel-reservation/cancel-reservation.service.ts:82]: What is this function doing? It just defines 2 local variables.

File: new-chatbot-app/src/library-api/libcal-api-auth/libcal-api-auth.service.ts [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+                  client_id: this.CLIENT_ID,
+                  client_secret: this.CLIENT_SECRET
+                }
+                // auth: {
+                //   username: this.CLIENT_ID,
+                //   password: this.CLIENT_SECRET,
+                // },
+              });
+              resolve(axiosResponse);
+            } catch (error: any) {
```
Comment (dominhnhut01) [inline new-chatbot-app/src/library-api/libcal-api-auth/libcal-api-auth.service.ts:62]: I suggest using HttpModule from Nest.js instead of straight axios. Also, I suggest using Observable.
You can look at this for reference: https://github.com/Meng-V/smartchatbot/blob/dev/new-chatbot-app/src/library-api/libapps-authorization/libapps-authorization.service.ts
They're pretty similar, only difference is the URL to request.

File: new-chatbot-app/src/llm-chain/llm-toolbox/libcal-tools/cancel-reservation/cancel-reservation.service.ts (hunk 5 above)
Comment (dominhnhut01) [reply new-chatbot-app/src/llm-chain/llm-toolbox/libcal-tools/cancel-reservation/cancel-reservation.service.ts:82]: Also, don't request access token at the start of this function. It should be requested inside the constructor instead. I know I request it in this function in the legacy code but let's change it this way.
Reason: A new access token would be generated for every single call. That's unnecessary and has a lot of overhead

File: new-chatbot-app/src/llm-chain/llm-toolbox/libcal-tools/room-reservation/room-reservation.service.spec.ts [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+      providers: [RoomReservationService],
+    }).compile();
+
+    service = module.get<RoomReservationService>(RoomReservationService);
+  });
+
+  it('should be defined', () => {
+    expect(service).toBeDefined();
+  });
+});
```
Comment (dominhnhut01) [inline new-chatbot-app/src/llm-chain/llm-toolbox/libcal-tools/room-reservation/room-reservation.service.spec.ts:18]: I think you can delete this one since you haven't coded it. We will add it in future PRs.

File: new-chatbot-app/src/llm-chain/llm-toolbox/libcal-tools/room-reservation/room-reservation.service.ts [hunk 8]
```diff
@@ -0,0 +1,4 @@
+import { Injectable } from '@nestjs/common';
+
+@Injectable()
+export class RoomReservationService {}
```
Comment (dominhnhut01) [inline new-chatbot-app/src/llm-chain/llm-toolbox/libcal-tools/room-reservation/room-reservation.service.ts:4]: I think you can delete this one since you haven't coded it. We will add it in future PRs.

File: new-chatbot-app/src/llm-chain/llm-toolbox/libcal-tools/cancel-reservation/cancel-reservation.service.ts [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+  public readonly description: string =
+    "This tool is for cancelling study room reservation. Use Final Answer instead if you don't have enough required parameters yet. Don't include any single quotes in the paramter.";
+
+  public readonly parameters: { [parameterName: string]: string } = {
+    bookingID: "string [REQUIRED]"
+  };
+
+  /**
+   * Singleton instance of the CancelReservationService.
+   */
```
Comment (dominhnhut01) [inline new-chatbot-app/src/llm-chain/llm-toolbox/libcal-tools/cancel-reservation/cancel-reservation.service.ts:18]: This docstring is not accurate anymore. I suggest deleting it

File: new-chatbot-app/src/llm-chain/llm-toolbox/libcal-tools/cancel-reservation/cancel-reservation.service.spec.ts [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+    const cancelResponse = { success: true };
+
+    mockLibcalApiAuthorizationService.getAccessToken.mockResolvedValue(accessToken);
+
+    const result = await service.serviceRun({ bookingID: bookingID });
+
+    const expectedAnswer = `Room reservation with ID: ${bookingID} is cancelled successfully\n`;
+
+    expect(mockLibcalApiAuthorizationService.getAccessToken).toHaveBeenCalled();
+    expect(result).toEqual(expectedAnswer);
```
Comment (dominhnhut01) [inline new-chatbot-app/src/llm-chain/llm-toolbox/libcal-tools/cancel-reservation/cancel-reservation.service.spec.ts:53]: I suggest mocking the HttpService to pretend that the cancel URL successfully and unsuccessfully cancels the room; therefore, test how the tool reacts to different scenario.

File: new-chatbot-app/src/llm-chain/llm-toolbox/libcal-tools/cancel-reservation/cancel-reservation.service.spec.ts [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+  });
+
+});
```
Comment (dominhnhut01) [inline new-chatbot-app/src/llm-chain/llm-toolbox/libcal-tools/cancel-reservation/cancel-reservation.service.spec.ts:56]: should also test the case when the input parameter to the tool is null or undefined. The LLM does that very often.

File: new-chatbot-app/src/llm-chain/llm-toolbox/llm-toolbox.module.ts [hunk 12]
```diff
@@ -1,10 +1,17 @@
...
 import { LibraryApiModule } from '../../library-api/library-api.module';
 import { HttpModule } from '@nestjs/axios';
+import { LibappsAuthorizationService } from 'src/library-api/libapps-authorization/libapps-authorization.service';
+import { RetrieveEnvironmentVariablesService } from 'src/shared/services/retrieve-environment-variables/retrieve-environment-variables.service';
 
 @Module({
   imports: [LibraryApiModule, HttpModule],
-  providers: [LibrarianSubjectLookupToolService],
+  providers: [
+    ConfigService,
```
Comment (dominhnhut01) [inline new-chatbot-app/src/llm-chain/llm-toolbox/llm-toolbox.module.ts:12]: ConfigService doesn't belong to LlmToolboxModule. It shouldn't be here

File: new-chatbot-app/src/llm-chain/llm-toolbox/llm-toolbox.module.ts [hunk 13]
```diff
@@ -1,10 +1,17 @@
...
+    LibrarianSubjectLookupToolService,
+    LibappsAuthorizationService,
+    RetrieveEnvironmentVariablesService],
```
Comment (dominhnhut01) [inline new-chatbot-app/src/llm-chain/llm-toolbox/llm-toolbox.module.ts:15]: LibappsAuthorizationService doesn't belong here in LlmToolboxModule. If you want to use it, imports: [LibraryApiModule] is enough. Same for RetrieveEnvironmentVariablesService, you need to import SharedModule instead

**Review by dominhnhut01 — CHANGES_REQUESTED** [review]
I left some comments/

File: new-chatbot-app/src/llm-chain/llm-toolbox/llm-toolbox.module.ts (hunk 12 above)
Comment (TranHVLoc) [reply new-chatbot-app/src/llm-chain/llm-toolbox/llm-toolbox.module.ts:12]: Backend can't run if I don't import and use it

File: new-chatbot-app/src/llm-chain/llm-toolbox/llm-toolbox.module.ts (hunk 13 above)
Comment (TranHVLoc) [reply new-chatbot-app/src/llm-chain/llm-toolbox/llm-toolbox.module.ts:15]: Backend can't run if I don't import and use it

File: new-chatbot-app/prisma/schema.prisma [hunk 14]
```diff
@@ -0,0 +1,55 @@
+// This is your Prisma schema file,
```
Comment (dominhnhut01) [inline new-chatbot-app/prisma/schema.prisma:1]: I think we should remove these prisma for now. Because we haven't touched the database in this CancelReservationTool. Let's put it in for future PRs.

File: new-chatbot-app/src/llm-chain/llm-toolbox/llm-toolbox.module.ts (hunk 13 above)
Comment (dominhnhut01) [reply new-chatbot-app/src/llm-chain/llm-toolbox/llm-toolbox.module.ts:15]: Because you didn't export the libapps-auth-service in the libcal-api module
